source/draw/Canvas.ooc

```diff
@@ -0,0 +1,68 @@
+//
+// Copyright (c) 2011-2014 Simon Mika <simon@mika.se>
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+use ooc-base
+use ooc-math
+use ooc-collections
+import Image, Pen
+
+Canvas: abstract class {
+	_size: IntSize2D
+	size ::= this _size
+	_pen := Pen new()
+	pen: Pen { get { this _pen } set(value) { this _pen = value } }
+	viewport: IntBox2D { get set }
+	blend: Bool { get set }
+	opacity: Float { get set }
+	_transform := FloatTransform3D identity
+	transform: FloatTransform3D { get { this _transform } set(value) { this _transform = value } }
+	focalLength: Float { get set }
+	init: func (=_size) {
+		this viewport = IntBox2D new(size)
+		this focalLength = 0.0f
+		this blend = false
+		this opacity = 1.0f
+	}
+	drawPoint: virtual func (position: FloatPoint2D) {
+		list := VectorList<FloatPoint2D> new()
+		list add(position)
+		this drawPoints(list)
+		list free()
```

Comment:
The thing is, on a GPU the only implementation needed is `drawPoints` since they are always batched. Therefore `drawPoint` is a special case of `drawPoints`. In the CPU implementation, the opposite is true. Therefore it is not obvious which of `drawPoint` and `drawPoints` can be generalized in the abstract class. I just picked one as default and then it's up to the implementation to override if it can be optimized.